File: current_session/78.py
"""
https://leetcode.com/problems/subsets/
"""
class Solution(object):
    def subsets(self, nums):
        """
        :type nums: List[int]
        :rtype: List[List[int]]
        """
        # 2**L
        # use binary
        c = 2**len(nums)
        ans = []

        for i in range(c):
            arr, tmp = bin(i)[2:].zfill(len(nums)), []
            for i in range(len(arr)):
                if arr[i] == '1':
                    tmp += nums[i],
            ans += tmp,

        return ans

File: current_session/c++/78.cpp
#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
  vector<vector<int>> res;
  vector<vector<int>> subsets(vector<int>& nums) {
    helper(nums, vector<int>{}, 0);
    return res;
  }
  void helper(vector<int>& nums, vector<int> tmp, int start) {
    res.push_back(tmp);
    for (int i=start; i<nums.size(); ++i) {
      tmp.push_back(nums[i]);
      helper(nums, tmp, i+1);
      tmp.pop_back();
    }
  }
};